### PR TITLE
[Sage-212] Form Select - Foundations Update

### PR DESCRIPTION
**Pull request by anechol**
## Description

Updates Form Select component:
- Adjusts margin for labels to `margin-bottom: 6px`
- Moves label markup above select input in both React and Rails
- Removes floating label
- Removes icon from help messages
- Adds `label` prop to map to label text
- Adds `id` prop to map to `name` and `id` attributes
- Updates color

### Implementation Changes
Existing instances of this component, including simpleform, will need to be updated to accommodate the restructured [markup](https://kajabi.atlassian.net/browse/SAGE-319).

[Figma](https://www.figma.com/file/sMbtLUHSt2vfKgKjyQ3052/%E2%9A%A0%EF%B8%8F-%5BDO-NOT-USE-WIP%5D-Sage-Update?node-id=430%3A3530)

### Follow-Up Items
- [Add Icon options](https://kajabi.atlassian.net/browse/SAGE-213)
- [Refactor `name` in KP to use `label`](https://kajabi.atlassian.net/browse/SAGE-214)
- [Arrow icon should be flipped when menu is open](https://kajabi.atlassian.net/browse/SAGE-335)
- [Update error state CSS classnames](https://kajabi.atlassian.net/browse/SAGE-337)

## Screenshots

|  Before  |  After  |
|--------|--------|
Screen Shot 2022-02-28 at 11 43 37 AM|Screen Shot 2022-02-28 at 11 33 44 AM

## Testing in `sage-lib`

- View [Rails component](http://localhost:4000/pages/component/form_select)
- Check that component is properly updated to design specs.
- Check that id and name props function as intended.

### React
- View [Storybook](http://localhost:4100/?path=/docs/sage-select--default)
- Check that component is properly updated to design specs.

## Testing in `kajabi-products`

(**BREAKING**) Updates Form Select component with new prop and UI changes. Existing instances of this component, including simpleform, will need to be updated to accommodate the restructured markup.

## Related

[Sage-212](https://kajabi.atlassian.net/browse/SAGE-212)

### CODE REVIEW

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb [hunk 1]
```diff
@@ -7,6 +12,7 @@
   data-js-select="true"
   <%= component.generated_html_attributes.html_safe %>
 >
+  <label class="sage-select__label" for="<%= selectName %>"><%= label %></label>
```
Comment (pixelflips) [inline docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:15]: This is another instance where we will need to note that the markup structure has changed in relation to SimpleForm within KP. There are quite a few instances when doing a quick search for `sage-select__field` in KP

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss [hunk 2]
```diff
@@ -4,63 +4,38 @@
...
-    background-color: $-select-color-label-background;
-    transition: transform 0.15s ease-in-out, color 0.15s ease-out;
-  }
+  margin-bottom: $-select-margin-label;
 }
 
 .sage-select__arrow::before {
-  @include sage-icon-base(caret-down);
-  @include position(0, sage-spacing(sm), unset, unset);
+  @include sage-icon-base(down-small);
```
Comment (pixelflips) [inline packages/sage-assets/lib/stylesheets/components/_form_select.scss:32]: May be out of scope for these updates, but in the mocks, it appears the arrow flips up and down. Not a show-stopper, but might be worth noting for a follow-up.

**Review by pixelflips — APPROVED** [review]
LGTM! Left a few general comments, but non-blockers.

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb (hunk 1 above)
Comment (anechol) [reply docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:15]: Thanks for this callout! I'll add that to the description.

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss (hunk 2 above)
Comment (anechol) [reply packages/sage-assets/lib/stylesheets/components/_form_select.scss:32]: @pixelflips Just updated this. Should work as expected now.

Comment (philschanely) [reply packages/sage-assets/lib/stylesheets/components/_form_select.scss:32]: I'm not sure its meant to stay flipped when selected but only while the menu is open.

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb (hunk 1 above)
Comment (philschanely) [reply docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:15]: I believe the convention with variables in Rails is to use `snake_case`. Sorry to nitpick.

**Review by philschanely — CHANGES_REQUESTED** [review]
Looking really good overall. Two small details to circle back to.

**Comment by anechol** [comment]
> Looking really good overall. Two small details to circle back to.

@philschanely Thanks, I'll get these updated. Don't worry about nitpicking, I do want these little things pointed out. It's helpful!

**Comment by anechol** [comment]
Created a follow-up ticket in Jira to handle the change to the icon being flipped. It's more involved and out of scope for this ticket than expected. All other visual changes should be in place.

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss [hunk 3]
```diff
@@ -115,41 +82,38 @@ $-select-padding-label: sage-spacing(2xs);
...
+      color: sage-color(grey, 600);
     }
   }
 }
 
 .sage-select__message {
   @include sage-form-field-message;
+
+  + .sage-select__arrow::before {
+    @include position(31%, sage-spacing(sm), unset, unset);
```
Comment (monicawheeler) [inline packages/sage-assets/lib/stylesheets/components/_form_select.scss:109]: For this `31%` value and `41%` value on line 114, is it possible to represent this as a calculation within a variable? If not, could it be commented as to why it's written this way? Something like, "Percentage is hard-coded due to specific positioning needed within the input space."

**Review by monicawheeler — CHANGES_REQUESTED** [review]
Small item on magic numbers for the `.sage-select__arrow`

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss [hunk 4]
```diff
@@ -115,41 +82,38 @@ $-select-padding-label: sage-spacing(2xs);
     border-color: $-select-color-success;
     box-shadow: $-select-border-box-shadow-size $-select-color-success;
 
-    .sage-select--error & {
+    .sage-form-field--error & {
```
Comment (monkeypox8) [inline packages/sage-assets/lib/stylesheets/components/_form_select.scss:85]: Added one more [follow-up item,](https://kajabi.atlassian.net/browse/SAGE-337) as renaming the error class here will impact a few crucial areas:

```
app/javascript/common/components/DynamicSelect/DynamicSelect.jsx:
  31:       'sage-select--error': hasError,

app/views/admin/coupons/_form.html.erb:
  93:             <div class="sage-select<%= " sage-select--error" if coupon.errors[:currency].present? %>" data-js-select="true">

config/initializers/simple_form_sage.rb:
  65:   config.wrappers :sage_select, class: "sage-select", error_class: "sage-select--error", wrapper_html: { "data-js-select": true } do |b|
```

**Review by monkeypox8 — APPROVED** [review]
👍 LGTM, nice work!

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss (hunk 4 above)
Comment (anechol) [reply packages/sage-assets/lib/stylesheets/components/_form_select.scss:85]: Thanks, I've added this to the description.

File: packages/sage-assets/lib/stylesheets/components/_form_select.scss [hunk 5]
```diff
@@ -16,6 +16,11 @@ $-select-filled-top-padding: rem(4.4px);
 $-select-height: rem(40px);
 $-select-padding-x: sage-spacing(sm);
 $-select-margin-label: rem(6px);
+$-select-arrow-font-size: rem(24px);
+$-select-arrow-position: calc(100% - #{$-select-height});
```
Comment (monicawheeler) [inline packages/sage-assets/lib/stylesheets/components/_form_select.scss:20]: 🙌

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb [hunk 6]
```diff
@@ -1,4 +1,9 @@
-<% selectID = component.name.downcase.dasherize %>
+<% 
+  # Temporary variables here until refactor for 'name' is complete
```
Comment (monicawheeler) [inline docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:2]:
```suggestion
  # Temporary variables here until refactor for 'name' is complete; see https://kajabi.atlassian.net/browse/SAGE-214
```

Comment (monicawheeler) [reply docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:2]: @anechol - just a minor consideration here

**Review by monicawheeler — APPROVED** [review]

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb (hunk 6 above)
Comment (anechol) [reply docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb]: Added this in. Thanks!

**Pull request merged**